aiokafka/record/legacy_records.py

```diff
@@ -278,155 +285,154 @@ def __repr__(self):
-        if not (value is None or
-                isinstance(value, (bytes, bytearray, memoryview))):
-            raise TypeError(
-                "Not supported type for value: {}".format(type(value)))
-
-        # Check if we have room for another message
-        pos = len(self._buffer)
-        size = self.size_in_bytes(offset, timestamp, key, value)
-        # We always allow at least one record to be appended
-        if offset != 0 and pos + size >= self._batch_size:
-            return None
-
-        # Allocate proper buffer length
-        self._buffer.extend(bytearray(size))
 
-        # Encode message
-        crc = self._encode_msg(pos, offset, timestamp, key, value)
+        # calculating length is not cheap; only do it once
+        key_size = len(key) if key is not None else 0
+        value_size = len(value) if value is not None else 0
 
-        return LegacyRecordMetadata(offset, crc, size, timestamp)
-
-    def _encode_msg(self, start_pos, offset, timestamp, key, value,
-                    attributes=0):
+        pos = self._pos
+        size = self.size_in_bytes(offset, timestamp, key_size, value_size)
+        # always allow at least one record to be appended
+        if offset != 0 and pos + size >= self._batch_size:
+            return None
+        elif offset == 0 and size > self._batch_size:
+            # if adding the message fails for any reason, this buffer will
+            # remain in its expanded state. Since position is tracked
+            # independently and comparisons are made against self._batch_size,
+            # this will have no ill effects
+            self._buffer = bytearray(size)
+
+        try:
+            crc = self._encode_msg(
+                pos, size, offset, timestamp, key_size, key, value_size, value)
+            self._pos += size
+            return LegacyRecordMetadata(offset, crc, size, timestamp)
+
+        except struct.error:
+            # perform expensive type checking only to translate struct errors
+            # to human-readable messages
+            if type(offset) != int:
+                raise TypeError(offset)
+            if type(timestamp) != int:
+                raise TypeError(timestamp)
+            if not isinstance(key, (bytes, bytearray, memoryview, NoneType)):
+                raise TypeError("Unsupported type for key: %s" % type(key))
+            if not isinstance(value, (bytes, bytearray, memoryview, NoneType)):
+                raise TypeError("Unsupported type for value: %s" % type(value))
+            raise
+
+    def _encode_msg(self, start_pos, size, offset, timestamp, key_size, key,
+                    value_size, value, attributes=0):
         """ Encode msg data into the `msg_buffer`, which should be allocated
             to at least the size of this message.
         """
         magic = self._magic
-        buf = self._buffer
-        pos = start_pos
+        buf = memoryview(self._buffer)[start_pos:]
```

Comment:
Why do we create a memoryview here? We can set the position to write from in pack_into call.

Reply:
Three reasons, in order of ascending importance:
1. It removes start_pos from the pack_into calls which makes them a bit clearer.
2. It removes start_pos from the slice on line 382, making it clearer at a glance.
3. It makes the slice on line 382 free (in terms of memory). Slicing bytearrays requires creating new objects while slicing memoryviews does not.

Since creating the memoryview is extremely cheap, I figured the gains were worth it.